Bracket Matching Refactor

alanmcruickshank opened the pull request:
Another biggie. Resolves #1046, relates to #962.

This took the linting time of out large dbt project from 16:37 to 7:30. That's over 50% so I'm pretty pleased.

The bottom line of what I've done is make bracketed elements their own segment and then made sure that in most cases, even where matching isn't successful, that this parsing is persisted. This has the following effects:
- The long parsing steps at the start of any file where we're mapping out the overall structure are *much* faster. This is because we only do the bracket matching once, and then later searches through the file (using the bracket sensitive matching) can go much faster because they reuse the previous markings around bracketed regions. This is the core idea of this PR.
- It changes a lot of the dialect steps around brackets. Unfortunately this changes *a lot* of the test cases, grammar routines and rules. These all follow from the first point above, but it makes this a big PR.
- I've added a few small things at the same time while I was in those places in the codesbase (adding python implementation to the config output, updating parts of the Exasol dialect to use `OptionallyBracketed`, basic `pyproject.toml` file, moving some dialect tests into their own `.sql` files rather than being in the python test cases).

barrywhart commented:
So much chonk, ha ha! 🐈 😹  I will take a look.

alanmcruickshank commented:
> So much chonk, ha ha! 🐈 😹 I will take a look.

Yeah sorry.

With this one I wanted to make sure it would actually deliver the performance gains I thought it would before putting a PR together. Unfortunately, because bracket matching is so far down in the stack, it involved a lot of refactoring to get to that point.

barrywhart commented on src/sqlfluff/core/parser/grammar/sequence.py:
```diff
@@ -245,48 +251,57 @@ def match(
-            seg_buff = start_match.unmatched_segments
+        # Are we dealing with a pre-existing BracketSegment?
+        if seg_buff[0].is_type("bracketed"):
+            seg: BracketedSegment = seg_buff[0]  # type: ignore
```
```suggestion
            seg: BracketedSegment = cast(BracketedSegment, seg_buff[0])
```

barrywhart commented on src/sqlfluff/core/parser/grammar/sequence.py:
```diff
@@ -295,43 +310,59 @@ def match(
             if not self._elements or (
-                all(e.is_optional() for e in self._elements) and self.allow_gaps
+                all(e.is_optional() for e in self._elements)
+                and (self.allow_gaps or (not pre_nc and not post_nc))
```
I should probably know this, but what's `nc`?

alanmcruickshank replied:
ah, _non code_. I'll add some comments

alanmcruickshank replied:
Actually.... I'll just rename them to `pre_segs` and `post_segs` 👍

barrywhart commented on src/sqlfluff/core/parser/grammar/sequence.py:
```diff
@@ -295,43 +310,59 @@ def match(
+            # Have we already got indents?
+            meta_idx = None
+            for idx, seg in enumerate(bracket_segment.segments):
+                if seg.is_meta and seg.indent_val > 0:  # type: ignore
```
```suggestion
                if seg.is_meta and cast(MetaSegment, seg).indent_val > 0:
```

barrywhart commented on src/sqlfluff/core/parser/grammar/sequence.py:
```diff
@@ -295,43 +310,59 @@ def match(
+                if seg.is_meta and seg.indent_val > 0:  # type: ignore
+                    meta_idx = idx
+                    break
+            # If we've already go indents, don't add more.
```
```suggestion
            # If we've already got indents, don't add more.
```

barrywhart commented on src/sqlfluff/core/parser/segments/base.py:
```diff
@@ -1033,6 +1037,53 @@ def iter_patches(self, templated_str: str) -> Iterator[FixPatch]:
+    def __init__(
+        self,
+        *args,
+        start_bracket: Tuple[BaseSegment] = None,
```
Why is it a tuple rather than just a `BaseSegment`?

alanmcruickshank replied:
Good question. I did this initially so that `tuple` arithmetic was easy, but these should always be single segments 🤔. For someone reading the code raw, do you think `BaseSegment` would make more sense? If so I make change it to do that instead.

barrywhart replied:
I'm torn, because the syntax for creating a tuple "on the fly" is really obtuse and easy to overlook. Let's stick with the current code, but maybe add a clarifying comment that there'll be only one?

barrywhart commented on src/sqlfluff/core/parser/segments/ephemeral.py:
```diff
@@ -67,17 +69,22 @@ def wrapped_match_method(self, segments: tuple, parse_context):
-                        parse_grammar=new_grammar,
-                        name=self.ephemeral_name,
-                    ),
+            # We shouldn't allow nested ephemerals. If they're present. Don't create another.
```
```suggestion
            # We shouldn't allow nested ephemerals. If they're present, don't create another.
```

barrywhart commented on src/sqlfluff/dialects/dialect_ansi.py:
```diff
@@ -154,12 +155,16 @@
-        ("round", "StartBracketSegment", "EndBracketSegment"),
-        ("square", "StartSquareBracketSegment", "EndSquareBracketSegment"),
-        ("curly", "StartCurlyBracketSegment", "EndCurlyBracketSegment"),
+        ("round", "StartBracketSegment", "EndBracketSegment", True),
```
Nitpick: The comment above is helpful. Maybe someday use `typing.NamedTuple` as well, to make the code more self documenting?

alanmcruickshank replied:
Agreed. I'm going to leave it for now, but if it gets any more complicated we probably should.

barrywhart commented on src/sqlfluff/dialects/dialect_exasol.py:
```diff
@@ -335,9 +335,7 @@ class GroupByClauseSegment(BaseSegment):
-                Sequence(
-                    Ref("StartBracketSegment"), Ref("EndBracketSegment")
-                ),  # () possible
+                Bracketed(),  # () possible
```
The `() possible` comment is a bit cryptic. Does this mean "Empty parentheses are possible"?

alanmcruickshank replied:
It does. I'm going to change it. 👍

barrywhart commented on src/sqlfluff/dialects/dialect_exasol.py:
```diff
@@ -396,9 +394,7 @@ class GroupingSetsClauseSegment(BaseSegment):
-                Sequence(
-                    Ref("StartBracketSegment"), Ref("EndBracketSegment")
-                ),  # () possible
+                Bracketed(),  # () possible
```
Ditto

barrywhart commented on src/sqlfluff/dialects/dialect_exasol.py:
```diff
@@ -1000,9 +994,7 @@ class AlterTableAddColumnSegment(BaseSegment):
         "ADD",
         Ref.keyword("COLUMN", optional=True),
         Ref("IfNotExistsGrammar", optional=True),
-        Ref("StartBracketSegment", optional=True),
```
Heh, this old code was a little scary -- allowing unmatched brackets!!

barrywhart commented on src/sqlfluff/rules/L018.py:
```diff
@@ -81,13 +81,18 @@ def indent_size_up_to(segs):
             balance = 0
             with_indent, with_indent_str = indent_size_up_to(raw_stack_buff)
-            for seg in segment.iter_segments(expanding=["common_table_expression"]):
+            print("SEG", segment, with_indent, repr(with_indent_str))
```
Looks like debugging code -- should this `print()` call be removed?

barrywhart commented on src/sqlfluff/rules/L018.py:
```diff
@@ -81,13 +81,18 @@ def indent_size_up_to(segs):
+            for seg in segment.iter_segments(
+                expanding=["common_table_expression", "bracketed"], pass_through=True
+            ):
+                print(seg)
```
Ditto

barrywhart commented on src/sqlfluff/rules/L018.py:
```diff
@@ -81,13 +81,18 @@ def indent_size_up_to(segs):
                     balance -= 1
                     if balance == 0:
                         closing_bracket_indent, _ = indent_size_up_to(raw_stack_buff)
+                        print("INDENT", closing_bracket_indent)
```
Ditto

barrywhart commented on test/dialects/snowflake_test.py:
```diff
@@ -11,86 +11,10 @@
 @pytest.mark.parametrize(
     "segment_cls,raw",
     [
-        ("UseStatementSegment", 'USE ROLE "MY_ROLE";'),
```
Why remove these test cases?

alanmcruickshank replied:
I moved them to a file based test rather than being here. As per the note a few lines above, the `.sql`/`.yml` tests are more robust than these ones, unless they're testing parts of segments rather than a full file.

barrywhart commented:
@alanmcruickshank: Re "chonk", I am just joking. Great work on all the things recently! And farther back also, of course. 🙏

alanmcruickshank commented on src/sqlfluff/core/parser/grammar/base.py:
```diff
@@ -537,11 +552,32 @@ def _bracket_sensitive_look_ahead_match(
                                 # on.
+
+                                # Complete the bracketed info
+                                bracket_stack[-1].segments += (
```
I've been through the ones in this function and they're _similar_ but not the same. I think in this case I'll leave it as is.

alanmcruickshank commented on src/sqlfluff/core/parser/grammar/delimited.py:
```diff
@@ -184,13 +179,13 @@ def match(
                 # there's no sense to try matching
                 if self.min_delimiters and len(delimiters) < self.min_delimiters:
-                    return MatchResult.from_unmatched(segments)
+                    return MatchResult.from_unmatched(mutated_segments)
```
Good call - I'll see if I can make a test.

alanmcruickshank commented on src/sqlfluff/core/parser/grammar/base.py:
```diff
@@ -604,16 +641,26 @@ def _bracket_sensitive_look_ahead_match(
-                            return ((), MatchResult.from_unmatched(segments), None)
+                            # We return with the mutated segments so we can reuse any bracket matching.
+                            return (
+                                (),
```
Good call - I'll see if I can make a test.

alanmcruickshank commented on src/sqlfluff/core/parser/grammar/base.py:
```diff
@@ -604,16 +641,26 @@ def _bracket_sensitive_look_ahead_match(
-                        return ((), MatchResult.from_unmatched(segments), None)
+                        # unmatched exit. We return with the mutated segments so we can
+                        # reuse any bracket matching.
+                        return (
```
Good shout - I'll see what I can do...

alanmcruickshank replied:
Yep - I've managed to combine the _three_ instances of this code into a single code path. Great idea! 🎉

alanmcruickshank commented:
@barrywhart - I think I've addressed all of your feedback points, could I get another review?

Great idea on getting some better test coverage for some of the paths. I've added a few extra test cases, updated comments and rationalized the exit points from `_bracket_sensitive_look_ahead_match`. I've also added some type hints to `_bracket_sensitive_look_ahead_match` because I realise it was otherwise pretty opaque.

I'm going to be AFK for a few days, if you like the changes, would you mind merging?

barrywhart commented:
Sure! I'll do a final review tomorrow.
